drivers/input/keyboard/keyboard_gpio.c

```diff
@@ -0,0 +1,421 @@
+/*
+ * Copyright (c) 2022, tangchunhui@coros.com
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT keyboard_gpio
+
+/**
+ * @file
+ * @brief Keyboard-gpio driver
+ */
+
+#include <string.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/input.h>
+#include <zephyr/dt-bindings/input/input.h>
+
+#include "../input_internal.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(keyboard_gpio);
+
+#ifndef MAX
+#define MAX(x, y) (((x) > (y)) ? (x) : (y))
+#endif
+
+#define SCAN_INTERVAL	CONFIG_KEYBOARD_GPIO_SCAN_INTERVAL
+#define TIME_DEBOUNCE	CONFIG_KEYBOARD_GPIO_TIME_DEBOUNCE
+#define TIME_LONGPRESS	CONFIG_KEYBOARD_GPIO_TIME_LONGPRESS
+#define TIME_HOLDPRESS  CONFIG_KEYBOARD_GPIO_TIME_HOLDPRESS
+
+/**
+ * @brief Key information structure
+ *
+ * This structure gathers useful information about keyboard controller.
+ *
+ * @param label Key label.
+ * @param code  Key code.
+ * @param debounce_ms  Key debounce time
+ * @param longpress_ms Key long press time
+ * @param holdpress_ms Key hold press time
+ */
+struct key_info_dt_spec {
+	const char *label;
+	uint16_t code;
+	uint16_t debounce_ms;
+	uint16_t longpress_ms;
+	uint16_t holdpress_ms;
+};
+
+/**
+ * @brief Get KEY INFO DT SPEC
+ *
+ * @param node_id devicetree node identifier
+ */
+#define KEY_INFO_DT_SPEC_GET(node_id) \
+{ \
+	.label = DT_PROP(node_id, label), \
+	.code = DT_PROP(node_id, code), \
+	.debounce_ms = DT_PROP_OR(node_id, debounce_ms, 0), \
+	.longpress_ms = DT_PROP_OR(node_id, longpress_ms, 0), \
+	.holdpress_ms = DT_PROP_OR(node_id, holdpress_ms, 0), \
+}
+
+enum key_state_t {
+	KEY_STATE_NONE,
+	KEY_STATE_PRESSED,
+	KEY_STATE_LONG_PRESSED,
+	KEY_STAT_MAXNBR
+};
+
+enum key_event_t {
+	KEY_EVENT_RELEASE = KEY_RELEASE,
+	KEY_EVENT_PRESSED = KEY_PRESSED,
+	KEY_EVENT_LONG_PRESSED = KEY_LONG_PRESSED,
+	KEY_EVENT_HOLD_PRESSED = KEY_HOLD_PRESSED,
+	KEY_EVENT_LONG_RELEASE = KEY_LONG_RELEASE,
+	KEY_EVENT_NONE = 0xff,
+};
+
+struct kbd_gpio_driver {
```

Comment:
weird naming, the whole code in keyboard_gpio.c is the driver. 
Actually I don't see the point of not including of these attributes into kbd_gpio_data